File: Dynamic Programming/FrogJump.java
import java.io.*;
import java.util.*;

class FrogJump {
    public static int frogJump(int[] h, int[] dp, int n) {
        if (n == 0)
            return 0;
        if (dp[n] != -1)
            return dp[n];
        int left = frogJump(h, dp, n - 1) + Math.abs(h[n] - h[n - 1]);
        int right = Integer.MAX_VALUE;
        if (n > 1)
            right = frogJump(h, dp, n - 2) + Math.abs(h[n] - h[n - 2]);
        return dp[n] = Math.min(left, right);
    }

    public static void main(String[] args) {
        int n = 4;
        int[] h = new int[] { 10, 20, 30, 10 };
        // memorization
        int[] dp = new int[n + 1];
        Arrays.fill(dp, -1);
        System.out.println(frogJump(h, dp, n - 1));

        // Tabulation
        Arrays.fill(dp, -1);
        dp[0] = 0;
        for (int i = 1; i < n; i++) {
            int right = Integer.MAX_VALUE;
            int left = dp[i - 1] + Math.abs(h[i] - h[i - 1]);
            if (i > 1) {
                right = dp[i - 2] + Math.abs(h[i] - h[i - 2]);
            }
            dp[i] = Math.min(left, right);
        }
        System.out.println(dp[n - 1]);

        // Memory optimization
        int prev2 = 0, prev = 0;
        for (int i = 1; i < n; i++) {
            int right = Integer.MAX_VALUE;
            int left = prev + Math.abs(h[i] - h[i - 1]);
            if (i > 1) {
                right = prev2 + Math.abs(h[i] - h[i - 2]);
            }
            int curr = Math.min(left, right);
            prev2 = prev;
            prev = curr;
        }
        System.out.println(prev);
    }
}

File: Dynamic Programming/frogJump.cpp
#include<bits/stdc++.h>
using namespace std;

int find(vector<int>&step,int n,vector<int>&dp){
	int right = INT_MAX;
	if(n==0 )return 0;
	if(dp[n]!= -1) return dp[n];
	int left = find(step,n-1,dp) + abs(step[n] - step[n-1]);
	if(n>1) int right = find(step,n-2,dp) + abs(step[n] - step[n-2]);
	dp[n] = min(left,right);
	return dp[n];
}
int frogJump(vector<int> &step,int n){
//	vector<int> dp(n+1,-1);
//	return find(step,n-1,dp);
//	vector<int> dp(n,0);
//	dp[0] = 0;
//	for(int i = 1;i<n;i++){
//		int left = dp[i-1] + abs(step[i] - step[i-1]);
//		int right = INT_MAX;
//		if(i > 1)right = dp[i-2]+ abs(step[i] - step[i-2]);
//		dp[i] = min(left,right);
//	}
//	return dp[n-1];
	int prev = 0,prev2=0;
	for(int i = 1;i<n;i++){
		int left = prev + abs(step[i] - step[i-1]);
		int right = INT_MAX;
		right = prev+ abs(step[i] - step[i-2]);
		int cur = min(left,right);
		prev2 = prev;
		prev = cur;
	}
	return prev;
}
int main(){
	vector<int> step = {10,20,10,20,40,30,20};
	cout<<frogJump(step,step.size());
	return 0;
}
